model/task/task.go

```diff
@@ -2572,35 +2580,90 @@ func ArchiveMany(tasks []Task) error {
-		if err != nil {
-			return nil, err
+		var err error
+		if len(toArchive) > 0 {
+			oldTaskColl := evergreen.GetEnvironment().DB().Collection(OldCollection)
+			_, err = oldTaskColl.InsertMany(sessCtx, toArchive)
+			if err != nil {
+				return nil, errors.Wrap(err, "archiving tasks")
+			}
+		}
+		if len(taskIds) > 0 {
+			_, err = evergreen.GetEnvironment().DB().Collection(Collection).UpdateMany(sessCtx,
+				bson.D{{Key: "_id", Value: bson.D{{Key: "$in", Value: taskIds}}}}, // Query all display tasks and tasks
+				bson.A{ // Pipeline
+					bson.D{{Key: "$set", Value: bson.D{ // execution = execution + 1
+						{Key: ExecutionKey, Value: bson.D{
+							{Key: "$add", Value: bson.A{
+								"$" + ExecutionKey, 1,
+							}},
+						}},
+					}}},
+					bson.D{{Key: "$unset", Value: bson.A{
+						AbortedKey,
+						AbortInfoKey,
+						OverrideDependenciesKey,
+					}}}})
+			if err != nil {
+				return nil, errors.Wrap(err, "archiving tasks")
+			}
 		}
+		if len(execTasksIds) > 0 {
+			// Backwards compatibility call + LPE setting for all tasks
+			_, err = evergreen.GetEnvironment().DB().Collection(Collection).UpdateMany(sessCtx,
+				bson.D{{Key: "_id", Value: bson.D{{Key: "$in", Value: execTasksIds}}}}, // Query all execution tasks
+				bson.A{ // Pipeline
+					bson.D{{Key: "$set", Value: bson.D{ // Sets LatestParentExecution (LPE) = !exists(LPE) ? execution + 1 : LPE + 1
+						{Key: LatestParentExecutionKey, Value: bson.D{
+							{Key: "$cond", Value: bson.A{
+								bson.D{{Key: "$not", Value: bson.A{ // !exists(LPE)
+									"$" + LatestParentExecutionKey,
+								}}},
+								bson.D{{Key: "$add", Value: bson.A{ // execution + 1
+									"$" + ExecutionKey, 1,
+								}}},
+								bson.D{{Key: "$add", Value: bson.A{ // LPE + 1
+									"$" + LatestParentExecutionKey, 1,
+								}}},
+							}}},
+						}}},
+					}})
+
+			// Same call without backwards compatibility
+			// _, err = evergreen.GetEnvironment().DB().Collection(Collection).UpdateMany(sessCtx,
+			// 	bson.D{{Key: "_id", Value: bson.D{{Key: "$in", Value: toUpdateExecTaskIds}}}}, // Query all 'taskIds'
+			// 	bson.A{ // Pipeline
+			// 		bson.D{{Key: "$set", Value: bson.D{ // Sets LatestParentExecution (LPE) = LPE + 1
+			// 			{Key: LatestParentExecutionKey, Value: bson.D{{Key: "$add", Value: bson.A{
+			// 				"$" + LatestParentExecutionKey, 1,
+			// 			}}}}}},
+			// 		}})
 
-		taskColl := evergreen.GetEnvironment().DB().Collection(Collection)
-		_, err = taskColl.UpdateMany(ctx, bson.M{
-			IdKey: bson.M{
-				"$in": taskIds,
-			},
-		},
-			bson.M{
-				"$unset": bson.M{
-					AbortedKey:   "",
-					AbortInfoKey: "",
-				},
-				"$inc": bson.M{
-					ExecutionKey: 1,
-				},
-			})
-		if err != nil {
-			return nil, err
+			if err != nil {
+				return nil, errors.Wrap(err, "updating documents")
+			}
+
+			// Call to update all tasks that are actually restarting
+			_, err = evergreen.GetEnvironment().DB().Collection(Collection).UpdateMany(sessCtx,
```

Comment:
should be in an if toUpdateExecTaskIds > 0 right?

Reply:
I didn't include this, because if there are any execTasks, there must be at least one updating

Reply:
Users shouldn't be able to reset a display task that only has succeeding exec tasks-- but should I include it to cover all angles?

Reply:
Yes they can; this is the purpose of the general "restart when all is completed". It's similar to how you can restart a regular task that is successful.

Reply:
Will there be a 'restart failed when all is completed" (in 'restart when all is completed', execTasks == toUpdateExecTask)? And if there is a restarted failed when all is completed, if all execution tasks succeed, the display task succeeds, and would not be counted as failed and would not be restarted. If the display task is failed, one execution has failed at least, and execTask >= toUpdateExecTask > 0.

I think the use case in which execTask > 0 but toUpdateExecTask == 0 (which is what the if statement would prevent), is when a user goes to a display task, and clicks restart failed when finished (if it's a button) or restart failed. And in this case, I discussed a little bit with Bynn and we thought there should be something blocking the user form doing so because there are none failed (or in the case of the when finished button, it should abort restarting if all failed?).

I think it makes sense to add the if statement to protect from a DB error, invalidating the transaction, but I'm not sure if it would ever be used

Reply:
I tried setting the pipeline $unset to bson.M and I get

write exception: write errors: [$unset specification must be a string or an array]

Reply:
This is the current behavior of restarting a display task so we will not be adding a new button; you can play around with this in staging if you want. Right now if a user tries to restart a display task (using just the regular Restart button), we set ResetWhenFinished, which is checked during EndTask (i.e. we check if the last execution task in the display task is finished, and if we want to reset it when finished, and then we call TryResetTask). 

I'm not sure I understand the next part of your comment but it's important that we don't break this functionality. The thing we're adding in this project is to allow users to only restart failed if they want, as opposed to restarting all tasks, which is currently the only option.

Reply:
In the current behavior, execTask == toUpdateExecTask since all exec tasks always restart. I don't think there is a use case in which results in execTack > 0 but toUpdateExecTask == 0. If the display task is being archived, it will have at least one exec task that it will have to update (aka advance to next execution)

So the if toUpdateExecTask > 0 I _think_ is not needed but I might be overseeing the case in which it is, but adding the if statement might save a random use case that was unforeseeable. I'm not sure which action is the best, with or without the if statement